overlord/state/taskrunner.go

```diff
@@ -141,50 +185,70 @@ func (r *TaskRunner) Ensure() {
 		return
 	}
 
+	// Locks must be acquired in the same order everywhere.
 	r.state.Lock()
 	defer r.state.Unlock()
 
-	for _, chg := range r.state.Changes() {
-		if chg.Status() == DoneStatus {
+	for _, t := range r.state.Tasks() {
+		handlers, ok := r.handlers[t.Kind()]
+		if !ok {
+			// Handled by a different runner instance.
 			continue
 		}
-		logger.Debugf("Working on change %s: %s", chg.ID(), chg.summary)
-
-		tasks := chg.Tasks()
-		for _, t := range tasks {
-			// done or error are final, nothing to do
-			// TODO: actually for error progate to halted and their waited
-			status := t.Status()
-			if status == DoneStatus || status == ErrorStatus {
-				continue
-			}
 
-			// No handler for the given kind of task,
-			// this means another taskrunner is going
-			// to handle this task.
-			if _, ok := r.handlers[t.Kind()]; !ok {
-				continue
-			}
+		tb := r.tombs[t.ID()]
 
-			// we look at the Tomb instead of Status because
-			// a task can be in DoStatus even when it
-			// is not started yet (like when the daemon
-			// process restarts)
-			if _, ok := r.tombs[t.ID()]; ok {
+		if t.Status() == AbortStatus {
+			if tb != nil {
+				tb.Kill(nil)
 				continue
 			}
+			r.swapAbort(t)
```

Comment:
I wonder if maybe a name like `tryUndo(t)` might fit here as it feels more descriptive?